### PR TITLE
[EDA] Prevent deletion of a Term Record when there is a Related Child Term Record

### LINKED ISSUE #1124: Prevent deletion of a Term Record when there is a Related Child Term Record
Note: FOR INTERNAL USE ONLY. For support questions, please use the EDA Chatter group in The Power of Us Hub.
For feature requests, please use the Ideas tab in The Power of Us Hub.

Do you want to request a feature or report a bug?
Feature

What is the current behavior?
When users delete a TERM with TERM GRADES as child records, the system will allow users to delete the TERM and it will not delete child records if Prevent_Term_Deletion__c is not enabled.

What is the expected behavior?
We want to provide customers the ability to prevent deletion of TERMs if it has Term child records by allowing them to toggle a checkbox in the Hierarchy Settings.

What is the motivation / use case for changing the behavior?
By giving customers the ability to control whether a TERM with child records should be deleted or not, it will help them with maintaining a clean system.

Acceptance Criteria

Given: If Prevent_Term_Deletion__c checkbox is enabled in Hierarchy Settings
When: I try to delete a Term record with child records
Then: I should not be able to delete the Term record.

### PR DESCRIPTION
**Pull request by skosgi1**
# Critical Changes

# Changes
## Update to Ensure Data Integrity
To ensure greater data integrity, we've added functionality to prevent the deletion of Term records that have related Term records. This change is for internal-use only and is disabled at this time. You won't see any difference in functionality.
**Note:** We strongly recommend that EDA Admins don't manage Hierarchy Custom Settings directly and, instead, use the EDA Settings UI.

# Issues Closed
Closes #1124 

# New Metadata

# Deleted Metadata

# Testing Notes
Scenario 1: “Prevent Term Deletion” checkbox is enabled and Term record have related Term Grade, Course Connection, or Course Offering Schedule or Term records.
1) Check “Prevent Term Deletion” checkbox in Hierarchy Settings.
2) Create 2 Term records
3) Associate 1 Term record to another.
4) Delete the parent Term record.
5) You should receive a pop up error that you cannot delete this Term. See https://salesforce.quip.com/fXSbAE7hHWV8#LPKACAsjDGR for the error message.

Scenario 2: “Prevent Term Deletion” checkbox is disabled and Term record have related Term Grade, Course Connection, or Course Offering Schedule or Term records.
1) UnCheck “Prevent Term Deletion” checkbox in Hierarchy Settings.
2) Create 2 Term records
3) Associate 1 Term record to another.
4) Delete the parent Term record.
5) The Parent Term should be deleted

Scenario 3: “Prevent Term Deletion” checkbox is disabled and Term record have related Term Grade, Course Connection, or Course Offering Schedule or Term records.
1) Uncheck “Prevent Term Deletion” checkbox in Hierarchy Settings.
2) Create 2 Term record
3) Associate 1 Term record to another.
4) Delete the Term record
5) You should be able to delete the Parent Term record

Scenario 4: “Prevent Term Deletion” checkbox is disabled and Term record does not have related Term Grade, Course Connection, or Course Offering Schedule or Term records.
UnCheck “Prevent Term Deletion” checkbox in Hierarchy Settings.
1) Create a Term record
2) Delete the Term record
3) You should be able to delete the Parent Term record

### CODE REVIEW

File: src/classes/TERM_CannotDelete_TEST.cls [hunk 1]
```diff
@@ -39,237 +39,282 @@ private class TERM_CannotDelete_TEST {
...
+        insert term2;
+       
+       // Ensure there are 2 terms
+        List<Term__c> termsBeforeDelete = [SELECT Id FROM TERM__c];
+        System.assertEquals(2, termsBeforeDelete.size());
+
+        // Ensure the Parent Term is not deleted
+        try {
+            Test.startTest();
+                List<Term__c> termsToDelete = [SELECT Id FROM TERM__c ORDER BY Name ASC];
```
Comment (lifewithryan) [inline src/classes/TERM_CannotDelete_TEST.cls]: @skosgi1  -- Let's check and see if we can actually do ORDER BY Name here without breaking shield. I don't believe we can.

Comment (lifewithryan) [reply src/classes/TERM_CannotDelete_TEST.cls]: looks like we cannot: https://help.salesforce.com/articleView?id=security_pe_considerations_general.htm&type=5

**Comment by skosgi1** [comment]
> Looks like there's at least one place in the unit test where we are doing an ORDER BY Name -- Name is a potentially encryptable field. If a customer is using shield and chosen to encrypt the Name (yes that's an edge case, but) then it will fail for them.

@lifewithryan Sorry I should've know! I slightly modified a couple of methods to go around it. Can you please give it another shot? Thx.

File: src/classes/TERM_CannotDelete_TEST.cls (hunk 1 above)
Comment (skosgi1) [reply src/classes/TERM_CannotDelete_TEST.cls]: Fixed in the latest commit. Resolving convo.

**Comment by spinkelman** [comment]
Passes QA - Waiting to merge until we release 1.94
Ticket will be "In Waiting"

**Comment by spinkelman** [comment]
Did a quick smoke test and everything looked go: Merging and Closing
